Merge current version to main

Kriplingo opened the pull request.

Kriplingo committed:
Updated requirements.txt

Kriplingo committed:
Added initial elements to the UI

Kriplingo committed:
Styling changes

Kriplingo committed:
Added load and save configuration options

Kriplingo committed:
Updated poetry to include clinguin

Kriplingo committed:
Fixed requirements.txt deletion

Kriplingo committed:
Added button classes

Kriplingo committed:
Changed window layout

Kriplingo committed:
Added styling

Kriplingo committed:
Added back saving and loading buttons

Kriplingo committed:
Updated poetry.lock

Kriplingo committed:
Added menu icon

Kriplingo committed:
Removed show statements in ooasp_simple.lp

Kriplingo committed:
Divided node clingraph encoding

Kriplingo committed:
Started custom backend implementation

Kriplingo committed:
Removed fclingo from dependencies (temporary bugfix)

susuhahnml committed:
Fix clingraph, cleanup and UI update

Kriplingo committed:
Added solving suggestions and browsing functionality

susuhahnml committed:
Imporved clingraph and some fixes

Kriplingo committed:
Updated enforcement button

Kriplingo committed:
Added function call skeleton

susuhahnml committed:
UI update

Kriplingo committed:
Removed 'force_restart' from backend

Kriplingo committed:
Updated UI layout and added restart confirmation

Kriplingo committed:
Added object selection

Kriplingo committed:
Added help modal

Kriplingo committed:
Edited styling

Kriplingo committed:
Re-defined next solution button

Kriplingo committed:
Updated button label

susuhahnml committed:
UI changes from last coding

susuhahnml committed:
version

susuhahnml committed:
Performance fixes and some edits

susuhahnml committed:
Pimped UI

Kriplingo committed:
Divided ui code into component files

For readability

susuhahnml committed:
Help and show association

susuhahnml committed:
Load file option

Kriplingo committed:
Added load modal

susuhahnml committed:
Fixed double assoc

susuhahnml committed:
Improved CV violations

susuhahnml committed:
Adjustments to make global smart functions work with objects without class

susuhahnml committed:
Fixes to handle objects but small issue with redefinition

susuhahnml committed:
Checking performance issue

susuhahnml committed:
Include program directive. Option to remove objects form configuration

susuhahnml committed:
Clear buton. Fixed required objects button

susuhahnml committed:
Update ooasp/encodings/ooasp_guess.lp

Co-authored-by: Richard Comploi-Taupe <richard.taupe@siemens.com>

susuhahnml committed:
Removed files

Kriplingo committed:
Merge pull request #60 from siemens/object-bounds

Object bounds

nrueh commented on ooasp/encodings/ooasp_aux_kb.lp:
```diff
@@ -20,7 +20,7 @@ ooasp_attr_enum(C,N,"true") :-
     ooasp_attr(C,N,"boolean").
 
-% true, if attribute has domain
+% true, if attribute has dospecimain
```
Is this a typo?

Kriplingo replied:
Most likely :D (Fixed!)

nrueh commented on ooasp/smart_ooasp.py:
```diff
@@ -54,8 +54,7 @@ def __init__(
         self.next_id = 1
-        self.associations = []
-        self.object_atoms = []
+        self.assumptions = set()
```
Why a set?

Kriplingo replied:
This was a change done in order to avoid duplicit entries that caused issues when interacting.

Kriplingo replied:
@susuhahnml would know more details.

nrueh replied:
Ok, makes sense!

nrueh commented:
@Kriplingo done!

rtaupe commented on clinguin_ui/ooasp_backend.py:
```diff
@@ -0,0 +1,106 @@
+            ],
+        )
+        self.smart_solver.load_base()
+        # self.smart_solver.create_initial_objects()
```
Please don't include commented-out code

Kriplingo replied:
I will recheck all the files for commented out code.

rtaupe commented on clinguin_ui/ooasp_backend.py:
```diff
@@ -0,0 +1,106 @@
+        return super()._assumption_list.union(self.smart_solver.assumption_list)
+
+    def _prepare(self):
+        # self.smart_solver.ctl.assign_external(Function("check_potential_cv"), False)
```
Please don't include commented-out code

rtaupe commented on clinguin_ui/ui.lp:
```diff
@@ -0,0 +1,58 @@
+
+    % ================= Context menu
+
+% elem(edit_context(ID),context_menu,w) :- ooasp_domain(_,ID).
+%     elem(selector(ID),button,edit_context(ID)) :- ooasp_domain(_,ID).
+%     attr(selector(ID),label,"Select for editing") :- ooasp_domain(_,ID).
+%     attr(selector(ID),class,("bg-success")) :- ooasp_domain(_,ID).
+
+    % when(selector(ID),click,update,(selected_object,label,@format("{}",ID))) :- ooasp_domain(_,ID).
+    % when(selector(ID),click,update,(assoc_edit_options(ID),visibility,shown)) :- ooasp_domain(_,ID).
+    % when(selector(ID),click,update,(value_edit_options(ID),visibility,shown)) :- ooasp_domain(_,ID).
+    %when(selector(ID),click,update,(selector(ID),label,"Deselect")) :- ooasp_domain(_,ID).
+    %when(selector(ID),click,update,(selector(ID),class,("bg-warning"))) :- ooasp_domain(_,ID).
+
+    %elem(attr_context_menu(ID),context_menu, w) :- ooasp_isa(_,ID).
+
+        %elem(attribute_context(ID),button, edit_context(ID)) :- ooasp_isa(_,ID).
+        %attr(attribute_context(ID),label, "Set Attributes") :- ooasp_isa(_,ID).
+        %when(attribute_context(ID),click,update,(attr_context_menu(ID),visibility, shown)) :- ooasp_isa(_,ID).
+
+
+
+
+
+    % isa_remove(ID) :- _clinguin_assume(ooasp_isa(C, ID),true).
+    % elem(remove_isa(ID),button,edit_context(ID)) :- isa_remove(ID).
+    % attr(remove_isa(ID),icon,"fa-trash") :- isa_remove(ID).
+    % attr(remove_isa(ID),label,"Un-assign class.") :- isa_remove(ID).
+    % attr(remove_isa(ID),class,("text-white";"bg-danger")) :- isa_remove(ID).
+    % when(remove_isa(ID),click,call,remove_assumption(ooasp_isa(C,ID))) :- isa_remove(ID), _clinguin_assume(ooasp_isa(C, ID),true).
+
+    % dev borers
+% attr(E,class,("border-1";"border";"border-dark")):- elem(E, container, _).
```
Please don't include commented-out code

rtaupe commented on clinguin_ui/ui_add_object.lp:
```diff
@@ -0,0 +1,106 @@
+
+                domain_size(S):- #count{1,ID:ooasp_domain(_,ID)} = S.
+                elem(domain_size_txt,label,info_container).
+                attr(domain_size_txt,label,@format("Domain Size: {}",DS)) :- domain_size(DS). %TODO fix sizing
```
What exactly is to be done here?

Kriplingo replied:
If I remember correctly this display issue was fixed already, it was just a styling issue.
I removed the comment.

rtaupe commented on clinguin_ui/ui_menu_bar.lp:
```diff
@@ -0,0 +1,183 @@
+#const help_message = @concat(
+    "<p><i>This UI allows to create a configuration</i></p>",
+    "<h5>Add objects</h5>",
+    "<p>In the add 'Add object' section one can add a new object to the system. Only leaf classes can be added as objects. Click on + to add the </p>",
```
Is the first "add" really correct in this text?

rtaupe commented on clinguin_ui/ui_menu_bar.lp:
```diff
@@ -0,0 +1,183 @@
+    "<h5>Add objects</h5>",
+    "<p>In the add 'Add object' section one can add a new object to the system. Only leaf classes can be added as objects. Click on + to add the </p>",
+    "<h5>Edit objects</h5>",
+    "<p>Click on the node in the configuration graph to edit the object. The lower right pannel will show the selected node and posible changes. </p>",
```
```suggestion
    "<p>Click on the node in the configuration graph to edit the object. The lower right panel will show the selected node and posible changes. </p>",
```

Kriplingo replied:
Fixed

rtaupe commented on clinguin_ui/ui_menu_bar.lp:
```diff
@@ -0,0 +1,183 @@
+    "<h5>Edit objects</h5>",
+    "<p>Click on the node in the configuration graph to edit the object. The lower right pannel will show the selected node and posible changes. </p>",
+    "<h5>Partial and full configuration</h5>",
+    "<p>The yellow button on the menu can toggle between partial and full configuration. In partial configuration objects might not have values for attributes and lowebounds might not be filled. In full configuration the system will enforce the complete configuration. </p>",
```
Please check all text elements for grammar and spelling, I don't want to deliver all the details on this ...

rtaupe commented on clinguin_ui/ui_menu_bar.lp:
```diff
@@ -0,0 +1,183 @@
+        elem(help_modal,modal,w).
+        attr(help_modal,title,"Help").
+        attr(help_modal,size,"xl").
+        attr(help_modal,class,("justify-content-start")). %Check if this actually does anything.
```
What does this comment mean?

Kriplingo replied:
It was just a note for myself that I forgot to remove.
Gone now.

rtaupe commented on clinguin_ui/ui_menu_bar.lp:
```diff
@@ -0,0 +1,183 @@
+attr(unsat_msg, type, "error"):-_clinguin_unsat.
+attr(unsat_msg, title, "Unsatisfiable"):-_clinguin_unsat.
+
+    % elem(check_btn,button,menu_bar).
+    % attr(check_btn,label,"Check").
+    % attr(check_btn,icon,"fa-magnifying-glass").
+    % attr(check_btn,class,("btn-outline-danger")).
```
Why is this commented-out?

Kriplingo replied:
We decided not to use it and do the checks automatically, it is just a leftover.
Now removed.

Kriplingo committed:
Merge branch 'main' into lb/clinguin

rtaupe commented on ooasp/encodings/ooasp_aux_config.lp:
```diff
@@ -7,6 +7,8 @@
 
 ooasp_domain(cls,new_object).
 
+ooasp_unused(new_object):- #false: ooasp_isa(C, new_object).
```
@susuhahnml please add a comment to explain this rule, it looks rather unusual :)

susuhahnml replied:
I removed it. It was for a feature of the UI that we did not use.

rtaupe replied:
Thank you.
Could you still briefly explain how it works, please? I've seen similar rules in other places in this project, but I've never seen rules like it before. Is the body similar to `not ooasp_isa(_, new_object)`, or does it mean something different?

rtaupe commented on ooasp/encodings/ooasp_guess.lp:
```diff
@@ -1,6 +1,17 @@
 % Copyright (c) 2022 Siemens AG Oesterreich
```
```suggestion
% Copyright (c) 2022-2024 Siemens AG Oesterreich
```
Please remember to update and add this also in other files (that are touched or newly created)

Kriplingo replied:
@rtaupe would you like me to do this in a separate commit?

rtaupe replied:
That's not necessary.

rtaupe commented on ooasp/encodings/ooasp_simple.lp:
```diff
@@ -21,10 +21,10 @@ check_permanent_cv. % Checks for permanent constraints apply
+% #show global_ub/3.
+% #show global_lb/3.
+% #show association_needed/4.
+% #show .
```
Why is this commented-out?

Kriplingo replied:
So that everything is shown in this version, but we have a note on what are the needed things. @susuhahnml please correct me if I am wrong.

susuhahnml replied:
This was addressed in Teams right? I see the last version working well.

rtaupe commented on ooasp/encodings/viz_config copy.lp:
What is this file?

Kriplingo replied:
That should not be committed, I will remove it.

rtaupe commented on ooasp/encodings/viz_config_nodes.lp:
```diff
@@ -0,0 +1,35 @@
+% domain_node(Class,Object_id) :- ooasp_domain(Class,Object_id), not ooasp_isa(_,Object_id).
+% isa_node(ClassIsA,Object_id) :- ooasp_domain(Class,Object_id), ooasp_isa(ClassIsA,Object_id), not ooasp_isa_leaf(_,Object_id).
+% leaf_node(Class,Object_id) :- ooasp_isa_leaf(Class,Object_id), ooasp_class(Class).
+% isa_smallest(SUPER,Object_id) :-
+%     ooasp_isa(SUPER,Object_id),
+% 	not ooasp_isa(SUB, Object_id): ooasp_subclass(SUB,SUPER).
+% ooasp_isa_leaf(C,Object_id) :-
+%     ooasp_isa(C,Object_id),
+% 	ooasp_leafclass(C).
+
+% any_node(Object_id):- domain_node(Class,Object_id).
+% any_node(Object_id):- leaf_node(Class,Object_id).
+% any_node(Object_id):- isa_node(Class,Object_id).
+
+
+% unused_node(Class,Object_id) :- ooasp_domain(Class,Object_id), not _all(ooasp_isa(_,Object_id)).
+% unused_node(Class,Object_id) :- ooasp_domain(Class,Object_id), _clinguin_browsing, not ooasp_isa(_,Object_id).
+
+% infered_class_node(Class,Object_id) :- ooasp_domain(Class,Object_id), not _all(ooasp_isa(_,Object_id)).
+% unused_node(Class,Object_id) :- ooasp_domain(Class,Object_id), _clinguin_browsing, not ooasp_isa(_,Object_id).
+
```
Why is this commented-out?

Kriplingo replied:
Must be another leftover from a previous version.
I will remove it.

rtaupe reviewed (changes requested):
Thank you for these contributions!
Please address my and Nico's comments.

Kriplingo committed:
Fixed problems caused by naming changes

Kriplingo committed:
Addressed PR review comments

rtaupe commented on clinguin_ui/ui_menu_bar.lp:
```diff
@@ -1,17 +1,20 @@
+    "<p>Click on the node in the configuration graph to edit the object. The lower right panel will show the selected node and posible changes. </p>",
     "<h5>Partial and full configuration</h5>",
-    "<p>The yellow button on the menu can toggle between partial and full configuration. In partial configuration objects might not have values for attributes and lowebounds might not be filled. In full configuration the system will enforce the complete configuration. </p>",
+    "<p>The yellow button on the menu can toggle between a partial and a full configuration. In a partial configuration objects might not have values for attributes and lowebounds might not be filled. In full configuration the system will enforce the complete configuration. </p>",
```
Please check spelling & grammar

rtaupe commented on clinguin_ui/ui_menu_bar.lp:
```diff
@@ -1,17 +1,20 @@
+    "<p>The yellow button on the menu can toggle between a partial and a full configuration. In a partial configuration objects might not have values for attributes and lowebounds might not be filled. In full configuration the system will enforce the complete configuration. </p>",
     "<h5>Find incrementally</h5>",
-    "<p>Click on the 'Find incrementally' button to allow the system to create all the required objects to complete de configuration. </p>",
+    "<p>Click on the 'Find incrementally' button to allow the system to create all the required objects amd to complete the configuration. </p>",
```
Please check spelling & grammar (in all text elements)

Kriplingo replied:
I went through this, hopefully I did not miss anything anymore.

rtaupe commented on ooasp/encodings/ooasp_smart_generation.lp:
```diff
@@ -50,8 +53,10 @@ assoc_needs_object(ID1, ASSOC, 1..CMIN - N, C2, SIDE, new_object):-
-global_ub(C2, 1..((NUM1 + MAX - 1) / MAX)-NUM2, new_object):-
+%   Here, an additional rack is needed since the global upperbound of frames for one rack is 8 but the actual
+%   number of frames is 9.
+global_ub_gap(C2,((NUM1 + MAX - 1) / MAX)-NUM2,new_object):-
```
Why were the rules in this file changed again?

Kriplingo replied:
Because they were in this form in the first PR into main, so they were changed as a part of the conflict resolution, I kept the version from main in this case, as I have assumed this is the newer one, but it would proabably be good if either @susuhahnml or @nrueh could take a look if this is correct.

susuhahnml replied:
The one with 1... is the correct one. The original version in main had a bug that I fixed with this 1..

susuhahnml replied:
I reincorporated this in my commit

Kriplingo committed:
Improving grammar and wording in UI help text

susuhahnml committed:
- Fixed --viz
- Fixed UI broken due to renaming
- Fixed smart functions range
- Removed logs
- Removed unused_node
